pytorch_lightning/trainer/supporters.py

```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
 
         return length
 
+    def _get_len_recursive(self, data):
+        if isinstance(data, Dataset):
+            return len(data)
+
+        elif isinstance(data, (float, int)):
+            return data
+
+        elif isinstance(data, Mapping):
+            if isinstance(list(data.values())[0], (Mapping, Sequence, Dataset, Iterable)):
+                return {k: self._get_len_recursive(v) for k, v in data.items()}
+            else:
+                return self._get_len(data)
+        elif isinstance(data, Sequence):
+            data = list(data)
+            if isinstance(data[0], (Mapping, Sequence, Dataset, Iterable)):
```

Comment:
same here.

Reply:
`any(isinstance(v, (Mapping, Sequence, Dataset, Iterable)) for v in data)` no ?

Reply:
I wanted to avoid iterating over the whole thing twice @tchaton